### PR TITLE
Better Role Watching

### PR DESCRIPTION
**Pull request by TurtIeSocks**
Clears Discord Member Sessions for:
- guildMemberRemove
- guildMemberUpdate

Currently based on overall map perms. Open to suggestions.
Immediately clears sessions if it detects that a user with map perms leaves the guild or has their perms removed. 

Credits: *Borrowed* from Poracle role manager

### CODE REVIEW

**Review by versx — APPROVED** [review]

File: src/services/events/guildMemberUpdate.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+const { clearOtherSessions } = require('../session-store.js');
+
+module.exports = async (client, oldPresence, newPresence) => {
+    let before = false;
+    let after = false;
+    const roleBefore = oldPresence.roles.cache.find((role) => client.config.discord.perms.map.roles.includes(role.id));
+    const roleAfter = newPresence.roles.cache.find((role) => client.config.discord.perms.map.roles.includes(role.id));
+    if (roleBefore) before = true;
+    if (roleAfter) after = true;
```
Comment (versx) [inline src/services/events/guildMemberUpdate.js:9]:
```suggestion
    let before = oldPresence.roles.cache.find((role) => client.config.discord.perms.map.roles.includes(role.id));
    let after = newPresence.roles.cache.find((role) => client.config.discord.perms.map.roles.includes(role.id));
```
Simplify check

**Review by clburlison — APPROVED** [review]
This works great for removing access from the map. 

As such I'm happy to get this merged in as it stands. However if someone's access gets elevated from basic access (think raids) to paid (pokemon, quests, etc) this currently doesn't solve that use case. Might be worth doing that in another PR as I'm not sure of a quick fix. @TurtIeSocks 's decision on how to move forward.

**Comment by TurtIeSocks** [comment]
> This works great for removing access from the map.
> 
> As such I'm happy to get this merged in as it stands. However if someone's access gets elevated from basic access (think raids) to paid (pokemon, quests, etc) this currently doesn't solve that use case. Might be worth doing that in another PR as I'm not sure of a quick fix. @TurtIeSocks 's decision on how to move forward.

Hmm could probably write up some more rules in the guildMemberUpdate file.

File: src/services/events/guildMemberUpdate.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+const { clearOtherSessions } = require('../session-store.js');
+
+module.exports = async (client, oldPresence, newPresence) => {
+    const roleBefore = oldPresence.roles.cache.find((role) => client.config.discord.perms.map.roles.includes(role.id));
+    const roleAfter = newPresence.roles.cache.find((role) => client.config.discord.perms.map.roles.includes(role.id));
+    try {
+        if (roleBefore && !roleAfter) {
```
Comment (clburlison) [inline src/services/events/guildMemberUpdate.js:7]:
```suggestion
    const rolesBefore = oldPresence.roles.cache
                .filter(x => BigInt(x.id).toString())
                .keyArray();
    const rolesAfter = newPresence.roles.cache
                .filter(x => BigInt(x.id).toString())
                .keyArray();
    const perms = Object.values(client.config.discord.perms)
                .map(x => x.roles)
                .reduce(x => [...x]);
    const roleDiff = rolesBefore
                 .filter(x => !rolesAfter.includes(x))
                 .concat(rolesAfter
                    .filter(x => !rolesBefore.includes(x))
                 );
    try {
        if (perms.includes(roleDiff[0])) {
```

**Review by clburlison — CHANGES_REQUESTED** [review]
Try this change. It will reduce the client.config.discord.perms.X.X.roles into a single array. Then anytime one of these roles is added or removed from a member the user session cache will be clear.

**Comment by TurtIeSocks** [comment]
@clburlison I think we should just merge 🤷

**Review by clburlison — APPROVED** [review]

**Pull request merged**